01-messaging.md

```diff
@@ -82,6 +83,91 @@ however, adding a 6-byte padding after the type field was considered
 wasteful: alignment may be achieved by decrypting the message into
 a buffer with 6-bytes of pre-padding.
 
+
+## Wire Type-Length-Value Format
+
+Throughout the protocol, we use a TLV (Type-Length-Value) format to allow for
+the backwards-compatible addition of new fields to existing message types. This
+format describes the TLV encoding specific to wire messages.
+
+A `tlv_record` represents a single field, encoded in the form:
+
+* [`varint`: `type`]
+* [`varint`: `length`]
+* [`length`: `value`]
+
+A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
+concatenation of the encoded `tlv_record`s. When used to extend existing
+messages, a `tlv_stream` is typically placed after all currently defined fields.
+
+The `type` is a varint encoded using the bitcoin CompactSize format. It
+functions as a message-specific, 64-bit identifier for the `tlv_record`
+determining how the contents of `value` should be decoded.
+
+The `length` is a varint encoded using the bitcoin CompactSize format
+signaling the size of `value` in bytes.
+
+The `value` depends entirely on the `type`. Each `tlv_stream` has its own 64-bit
```

Comment:
Could you detail that a bit?
Does that mean that a `tlv_stream` MUST have the following format?

- [`8`:`namespace`]
- [`varint`:`length`]
- TLV records

It seems incompatible with the current wire lightning message format that use a 2-byte `type` (unless we then encode the namespace in the `payload` part).

I thought the `namespace` would be detected somewhat implicitly. For example in onion records, the `namespace` would be the onion version, and in wire messages it would be the 2-byte top-level `type`.

Reply:
Yeah the namespace isn’t encoded, it’s implied from the context (wire message, onion payload+realm byte, etc).

Which parts do you think need clarification?

Reply:
Ok then great, I didn't understand why you explicitly said that it uses a 64-bit namespace, maybe that part should be removed?

Reply:
Types are now 64-bit unsigned integers because of the use of CompactSize, hence each message has a 64-bit namespace. I added it to clarify the range of possible values a `type` can assume, but maybe there’s a better way to say that?

Reply:
Gotcha I slightly misunderstood the sentence, I thought it meant a tlv stream namespace needed to be identified by a 64-bit identifier. With your explanation the sentence makes more sense.
I'll try to think of different wordings, but nothing better comes to mind right now...

Reply:
removed the mention of 64-bit namespace from `value`, since it seemed redundant after already being addressed in the description of `type`